headerbp/signing.go

```diff
@@ -0,0 +1,179 @@
+	// into a header value.
+	b.WriteString(strconv.Itoa(len(manifest)))
+	b.WriteRune(delimiter)
+	for _, entry := range manifest {
+		b.WriteString(entry.normalized)
+		b.WriteRune(':')
+		b.WriteString(getHeader(entry.original))
+		b.WriteRune(delimiter)
+	}
+}
+
+type signHeadersOptions func()
+
+// SignHeaders signs the given headers with the given signing secret using baseplate message signing. The
+// signature will be valid for 5 minutes.
+//
+// This can be used by middlewares clients that send requests to other services that are exposed to untrusted traffic to
+// sign the headers before sending them or by services that are setting up the initial headers to be propagated.
+func SignHeaders(
+	ctx context.Context,
+	signingSecret secrets.VersionedSecret,
+	headerNames []string,
+	getHeader func(string) string,
+	opts ...signHeadersOptions,
+) (string, error) {
+	b := getBuffer()
+	defer putBuffer(b)
+
+	concatHeaders(b, headerNames, getHeader, signatureVersionPrefix)
+	signature, err := signing.Sign(
+		signing.SignArgs{
+			Message:   b.Bytes(),
+			Secret:    signingSecret,
+			ExpiresIn: 5 * time.Minute,
+		})
+	if err != nil {
+		return "", fmt.Errorf("generating signature: %w", err)
+	}
+
+	b.Reset()
+	b.WriteString(signatureVersionPrefix)
+	b.WriteRune('.')
+	b.WriteString(signature)
+	return b.String(), nil
+}
+
+type signatureComponents struct {
+	version       int
+	versionPrefix string
+	signature     string
+}
+
+func extractVersion(signature string) (*signatureComponents, error) {
+	version, sig, ok := strings.Cut(signature, ".")
+	if !ok {
+		return nil, fmt.Errorf("no version found in signature")
+	}
+	versionInt, err := strconv.Atoi(version)
+	if err != nil {
+		return nil, fmt.Errorf("invalid version format %q: %w", version, err)
+	}
+	return &signatureComponents{
+		version:       versionInt,
+		versionPrefix: version,
+		signature:     sig,
+	}, nil
+}
+
+// VerifyHeaders verifies the signature of the given headers using the given verification secret. If the signature
+// is valid, it sets the signature on the context.
+//
+// This can be used by middlewares that receive requests from untrusted traffic to verify the headers before recording
+// them to be propagated.
+func VerifyHeaders(
+	ctx context.Context,
+	verificationSecret secrets.VersionedSecret,
+	signature string,
+	headerNames []string,
+	getHeader func(string) string,
+) (context.Context, error) {
+	components, err := extractVersion(signature)
+	if err != nil {
+		return ctx, fmt.Errorf("%w: %w", ErrInvalidSignatureVersion, err)
+	}
+	if components.version != 1 {
+		return ctx, fmt.Errorf("%w: unsupported version number %d", ErrInvalidSignatureVersion, components.version)
+	}
+
+	b := getBuffer()
+	defer putBuffer(b)
+	concatHeaders(b, headerNames, getHeader, components.versionPrefix)
+	if err := signing.Verify(b.Bytes(), components.signature, verificationSecret); err != nil {
```

Comment:
oh but you are also using `Bytes`, in that case I guess it makes sense to use `bytes.Buffer` with a pool.

(or you can use an interface to make `concatHeaders` accept both `*bytes.Buffer` and `*strings.Builder`, but  that's probably too overkill 😅)